### PR TITLE
Fix #4203: Added CI checks to validate various properties of color files

### LINKED ISSUE #4203: Add CI checks to validate various non-layout color files properties.
Add CI check to validate various color properties including :

- Ensuring no colors in component_colors.xml or color_palette.xml directly define colors
- Ensuring all colors in component_colors.xml only reference colors from color_palette.xml
- Ensuring that all colors from color_palette.xml only reference colors from color_defs.xml
- Ensuring that all colors in color_defs.xml are raw colors (e.g. with a hex code)
- Ensuring that none of the colors in component_colors.xml or color_palette.xml are raw colors (e.g. with a hex code)
- Ensuring that all color declarations in component_colors and color_palette ends with _color suffix
- Ensuring that all color declarations in color_defs should not have color keyword
- Ensuring that all color declarations follow snake_case naming convention

### PR DESCRIPTION
**Pull request by ayush0402**
## Explanation

Fixes #4203 

Add CI check to validate various color properties including :
Added regex checks for naming conventions for color declarations under scripts/assets

- [x] Ensuring all colors in component_colors.xml only reference colors from color_palette.xml
- [x] Ensuring that all colors from color_palette.xml only reference colors from color_defs.xml
- [x] Ensuring that all colors in color_defs.xml are raw colors (e.g. with a hex code)
- [x] Ensuring that none of the colors in component_colors.xml or color_palette.xml are raw colors (e.g. with a hex code)
- [x] Ensuring that all color declarations in component_colors and color_palette ends with _color suffix
- [x] Ensuring that all color declarations in color_defs should not have color keyword
- [x] Ensuring that all color declarations follow snake_case naming convention

## Essential Checklist

- [x] The PR title and explanation each start with "Fix #bugnum: " (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] Any changes to [scripts/assets](https://github.com/oppia/oppia-android/tree/develop/scripts/assets) files have their rationale included in the PR explanation.
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary code changes from Android Studio ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#undo-unnecessary-changes)).
- [x] The PR is made from a branch that's **not** called "develop" and is up-to-date with "develop".
- [x] The PR is **assigned** to the appropriate reviewers ([reference](https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section)).

### CODE REVIEW

**Comment by BenHenning** [comment]
@ayush0402 two things:
1. If this is in ongoing development, please set it to 'Draft' to make sure no one thinks that it should be in review
2. You can test the regex script locally with Bazel by running the same command that the actions job runs (which will probably be much faster than waiting for CI to run each time)

**Comment by ayush0402** [comment]
@BenHenning Sure, I will do it. Thanks for the suggestion for running regex script locally.

**Comment by ayush0402** [comment]
@BenHenning Hello, I have added the regex checks part which is responsible to check naming conventions and similar.

Now, for the check including ensuring all colors in `component_colors.xml` only reference `color_palette.xml` and other similar checks which actually require comparing contents of multiple files...
I am thinking to approach this by adding scripts similar to exisitng TODO scripts we have. We can use the RepositoryFile class to retrieve contents from the xml file and then use Regex to extract and store names of the colors being referenced in `component_colors.xml` and similarly store all declarations from `color_palette.xml` in another list. We can then check if all the colors stored in list 1 are also present in list 2.

Does this sound good? (I haven't tried it yet and am still investigating more about this, so I might be wrong about this.)
Or if you have any better approach please let me know.

UPDATE : I started working on this approach and have pushed a few incomplete changes.

**Comment by BenHenning** [comment]
> @BenHenning Hello, I have added the regex checks part which is responsible to check naming conventions and similar.
> 
> Now, for the check including ensuring all colors in `component_colors.xml` only reference `color_palette.xml` and other similar checks which actually require comparing contents of multiple files... I am thinking to approach this by adding scripts similar to exisitng TODO scripts we have. We can use the RepositoryFile class to retrieve contents from the xml file and then use Regex to extract and store names of the colors being referenced in `component_colors.xml` and similarly store all declarations from `color_palette.xml` in another list. We can then check if all the colors stored in list 1 are also present in list 2.
> 
> Does this sound good? (I haven't tried it yet and am still investigating more about this, so I might be wrong about this.) Or if you have any better approach please let me know.
> 
> UPDATE : I started working on this approach and have pushed a few incomplete changes.

@ayush0402 one thing you might be able to do is leverage the fact that we can establish naming requirements to simplify things. Consider:

```textproto
file_content_checks {
  file_path_regex: "app/src/main/res/values/component_colors.xml"
  prohibited_content_regex: "name=\"(?!component_color_).+\""
  failure_message: "All colors in component_colors must start with 'component_color_'."
}
file_content_checks {
  file_path_regex: "app/src/main/res/layout.*?/.+?\\.xml"
  prohibited_content_regex: "@color/(?!component_color_).+"
  failure_message: "Only colors from component_colors.xml may be used in layouts."
}
```

Might something like that work? It might be simpler to leverage the existing regex checker than building a whole new CI check.

**Comment by ayush0402** [comment]
@BenHenning Wouldn't this then require us to change the existing decided conventions for color files to include the file name in each declarations? Currently color names are not tied to the files they are declared in but rather the layout they are intended to be used in.

**Comment by BenHenning** [comment]
> @BenHenning Wouldn't this then require us to change the existing decided conventions for color files to include the file name in each declarations? Currently color names are not tied to the files they are declared in but rather the layout they are intended to be used in.

Apologies for the late reply @ayush0402. I didn't actually intend for the file names to be included in the color name; that's actually a coincidental side effect. Given that we're already changing the current convention (since there are now colors that are specifically _not_ associated with a certain screen), it seems sensible to establish conventions to easily identify where a color belongs (for location and mental organization).

@rt4914 do you have any thoughts on this?

**Comment by rt4914** [comment]
> > @BenHenning Wouldn't this then require us to change the existing decided conventions for color files to include the file name in each declarations? Currently color names are not tied to the files they are declared in but rather the layout they are intended to be used in.
> 
> Apologies for the late reply @ayush0402. I didn't actually intend for the file names to be included in the color name; that's actually a coincidental side effect. Given that we're already changing the current convention (since there are now colors that are specifically _not_ associated with a certain screen), it seems sensible to establish conventions to easily identify where a color belongs (for location and mental organization).
> 
> @rt4914 do you have any thoughts on this?

@ayush0402 @BenHenning This is what I have understood from above conversations. We will change the naming convention of colors further to ensure that are regex checks are easy to write and also as soon as someone reads that color it automatically gets mapped to correct file. 

If this is correct, it makes sense to me.

**Comment by ayush0402** [comment]
@rt4914 @BenHenning So, I should create a PR renaming : 
- every color in `component_colors.xml` renamed to `<component_color>_<current_name>`
- every color in `color_palette.xml` renamed to `<color_palette>_<current_name>`
- every color in `color_defs.xml` renamed to `<color_def>_<current_name>`

Is this correct?

**Comment by BenHenning** [comment]
> @rt4914 @BenHenning So, I should create a PR renaming :
> 
> * every color in `component_colors.xml` renamed to `<component_color>_<current_name>`
> * every color in `color_palette.xml` renamed to `<color_palette>_<current_name>`
> * every color in `color_defs.xml` renamed to `<color_def>_<current_name>`
> 
> Is this correct?

I think that's sensible. I'm definitely on board with the component_colors prefix, but am less sure of the other two. Nevertheless, we can start with those since they're consistent with the file names and always change them in the future if they are, for some reason, unclear.

**Comment by ayush0402** [comment]
@BenHenning PTAL

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [inline scripts/assets/file_content_validation_checks.textproto]: Nit: please remove blank newline for consistency.

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [inline scripts/assets/file_content_validation_checks.textproto:329]: Does the migration file also need to be added as an exemption, or is the plan to check this in after the migration is complete?

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [inline scripts/assets/file_content_validation_checks.textproto]: Should this use ``\\x`` instead?

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [inline scripts/assets/file_content_validation_checks.textproto]: Nit: error doesn't match pattern (``defs`` vs. ``def``).

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [inline scripts/assets/file_content_validation_checks.textproto]: For this & other error messages, I suggest including the full xml filename to maybe help clarify the errors a bit.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:1779]: I suggest updating the new test names for tests that are verifying a specific file to be specific to that file, e.g. ``testFileContent_colorPalette_...``.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]:
```suggestion
  fun testFileContent_doesNotStartWithColorPalette_fileContentIsNotCorrect() {
```

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]: Ditto here (capitalization) & in other new test names if others have the same issue.

Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]: Also, should this be for not starting with colorDef? Ditto for other test names (please make sure they reflect the thing being tested).

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @ayush0402. I think this looks really solid! Had a bunch of nits--PTAL.

Beyond that, is this actually all of the checks we want to add, or should there also be checks to make sure that color references in Java, Kotlin, and other XML files only ever references color_component colors, as well?

File: scripts/assets/file_content_validation_checks.textproto
Comment (ayush0402) [reply scripts/assets/file_content_validation_checks.textproto]: Done

File: scripts/assets/file_content_validation_checks.textproto
Comment (ayush0402) [reply scripts/assets/file_content_validation_checks.textproto]: I am not aware about `\\x` use but it does not seem to working as intended when I tried replacing it with `\\w`.

File: scripts/assets/file_content_validation_checks.textproto
Comment (ayush0402) [reply scripts/assets/file_content_validation_checks.textproto]: Done.

File: scripts/assets/file_content_validation_checks.textproto
Comment (ayush0402) [reply scripts/assets/file_content_validation_checks.textproto]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt
Comment (ayush0402) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]: Done.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt
Comment (ayush0402) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]: Done.

Comment (ayush0402) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]: Done

File: scripts/assets/file_content_validation_checks.textproto
Comment (ayush0402) [reply scripts/assets/file_content_validation_checks.textproto:329]: I was thinking to not consider migration file for tests at this moment as we are not expecting to add anything to it anyway. Can you elaborate more on what you mean?

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt
Comment (ayush0402) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:1779]: I did that for every but one test as renaming that function exceeds the 100 characters limit. Can you suggest something for it?

**Comment by ayush0402** [comment]
@BenHenning PTAL.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt
Comment (BenHenning) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:1779]: Ah, are you referring to ``testFileContent_doesNotReferenceColorFromColorPalette_fileContentIsNotCorrect``?

Perhaps:

``testFileContent_componentColors_includesNonPaletteColorReferences_fileContentIsNotCorrect``

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [reply scripts/assets/file_content_validation_checks.textproto]: Ah, well the potential issue that I see is ``\\x`` checks for hexadecimal vs. ``\\w`` which checks for any letter, I believe. Since color codes can only contain hex characters I think the pattern might be a bit more correct by using ``\\x``.

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [reply scripts/assets/file_content_validation_checks.textproto:329]: Ah, well the main consideration I'm having is whether this PR should also be introducing enforcement of component_colors in layout files (as that's the missing link--at the moment, layout files are allowed to reference anything). My original question came from me thinking about the fact that we need to be isolating references of each type of color, but that actually happens by default with the way you've approached it so long as layouts and Kotlin/Java code is limited to only referencing component colors.

Is the plan to introduce that support later on?

**Review by BenHenning — COMMENTED** [review]
Thanks @ayush0402. Just had a few follow-up comments--PTAL.

Also, it appears that there are now conflicts to resolve from the develop branch, too.

File: scripts/assets/file_content_validation_checks.textproto
Comment (ayush0402) [reply scripts/assets/file_content_validation_checks.textproto:329]: I think we can introduce that support later on as currently most of the codebase references colors directly from color_defs.xml. We can introduce this check when dark mode is completely implemented and every layout is using component_colors.xml.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt
Comment (ayush0402) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt:1779]: Done.

File: scripts/assets/file_content_validation_checks.textproto
Comment (ayush0402) [reply scripts/assets/file_content_validation_checks.textproto]: Using `\\x` actually throws this error.
```
92) testFileContent_componentColors_doesNotHaveHexColorValue_fileContentIsCorrect(org.oppia.android.scripts.regex.RegexPatternValidationCheckTest)
java.util.regex.PatternSyntaxException: Illegal hexadecimal escape sequence near index 5
.>#\x+<\/color>
     ^
        at java.base/java.util.regex.Pattern.error(Pattern.java:2028)
        at java.base/java.util.regex.Pattern.x(Pattern.java:3380)
        at java.base/java.util.regex.Pattern.escape(Pattern.java:2609)
        at java.base/java.util.regex.Pattern.atom(Pattern.java:2286)
        at java.base/java.util.regex.Pattern.sequence(Pattern.java:2210)
        at java.base/java.util.regex.Pattern.expr(Pattern.java:2069)
        at java.base/java.util.regex.Pattern.compile(Pattern.java:1783)
        at java.base/java.util.regex.Pattern.<init>(Pattern.java:1430)
        at java.base/java.util.regex.Pattern.compile(Pattern.java:1069)
        at kotlin.text.Regex.<init>(Regex.kt:89)
        at org.oppia.android.scripts.regex.MatchableFileContentCheck$Companion.createFrom(RegexPatternValidationCheck.kt:244)
        at org.oppia.android.scripts.regex.RegexPatternValidationCheckKt.main(RegexPatternValidationCheck.kt:45)
        at org.oppia.android.scripts.regex.RegexPatternValidationCheckTest.runScript(RegexPatternValidationCheckTest.kt:2164)
        at org.oppia.android.scripts.regex.RegexPatternValidationCheckTest.testFileContent_componentColors_doesNotHaveHexColorValue_fileContentIsCorrect(RegexPatternValidationCheckTest.kt:1877)

FAILURES!!!
Tests run: 92,  Failures: 92
```

**Comment by ayush0402** [comment]
@BenHenning PTAL. Also there are some new changes pushed to develop which don't follow the set rules for the colors in this PR and hence the static checks are failing.

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [reply scripts/assets/file_content_validation_checks.textproto:329]: That sounds fine to me, but does that mean #4203 isn't being fully resolved, or should we split the later work into a follow-up issue? I just want to make sure that nothing is lost here given that the later checks are sort of critical for getting the most value out of the checks introduced in this PR.

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [reply scripts/assets/file_content_validation_checks.textproto]: Ah, this was my bad. ``\\x`` is meant to be used with a specific hex sequence, e.g. ``\\xFF``.

Since we're using Java, extended character classes should be supported. Does ``\\p{XDigit}`` work?

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt
Comment (BenHenning) [inline scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]:
```suggestion
  fun testFileContent_componentColors_doesNotStartWithComponentColors_fileContentIsNotCorrect() {
```

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @ayush0402! Apologies for the very long delay in my review.

I think the PR overall looks quite good. I just had a few follow-up comments--PTAL.

File: scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt
Comment (ayush0402) [reply scripts/src/javatests/org/oppia/android/scripts/regex/RegexPatternValidationCheckTest.kt]: Done

**Comment by ayush0402** [comment]
@BenHenning Sorry for delay on this. 
I think we create another an issue for checks for layout xml files and that can be tackled in a separate PR.

And yes, `\\p{XDigit}` seems to be working fine.

**Comment by ayush0402** [comment]
> @BenHenning PTAL. Also there are some new changes pushed to develop which don't follow the set rules for the colors in this PR and hence the static checks are failing.

@BenHenning In case you missed this.

**Comment by ayush0402** [comment]
@BenHenning Also fixed the added colors which weren't following the correct convention.

@rt4914 PTAL.

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [reply scripts/assets/file_content_validation_checks.textproto:329]: Reminder to please address this.

File: scripts/assets/file_content_validation_checks.textproto
Comment (BenHenning) [reply scripts/assets/file_content_validation_checks.textproto]: Per your comment on the main thread, it seems this has been addressed.

**Review by BenHenning — COMMENTED** [review]
Thanks @ayush0402. I think that overall the PR LGTM, there's just one unresolved comment thread--PTAL.

**Comment by ayush0402** [comment]
@BenHenning If I am understanding your concern correctly, this is regarding having checks for layout files accessing only component_colors.xml only, right? 
I was thinking to tackle that in a separate PR as many layout files are currently using colors outside component_colors.xml. Wouldn't having the checks right now fail them anyway? 
Since layout files are still being updated for the new convention that can be dealt with separately, I will create a separate issue for that if you suggest so.

WDYT?

**Comment by BenHenning** [comment]
> @BenHenning If I am understanding your concern correctly, this is regarding having checks for layout files accessing only component_colors.xml only, right? I was thinking to tackle that in a separate PR as many layout files are currently using colors outside component_colors.xml. Wouldn't having the checks right now fail them anyway? Since layout files are still being updated for the new convention that can be dealt with separately, I will create a separate issue for that if you suggest so.
> 
> WDYT?

@ayush0402 I think filing a separate issue is completely reasonable. I see two possibilities to make sure the accounting of the remaining work is clear:
1. #4203 is renamed to focus on non-layout colors and a new issue is filed for that
2. #4203 isn't closed as a result of this PR since it includes the layout colors which this PR isn't yet addressing

(Also an FYI: you can more easily reply to old conversations via the 'conversations' drop-down via the PR's 'files' tab).

File: scripts/assets/file_content_validation_checks.textproto
Comment (ayush0402) [reply scripts/assets/file_content_validation_checks.textproto:329]: Replied on the main thread.

**Comment by ayush0402** [comment]
@BenHenning I renamed the existing issue to focus on non-layout files and also created a new issue for the layout files (#4396).
PTAL.

**Review by BenHenning — APPROVED** [review]
Thanks @ayush0402 for addressing the issues--everything looks really!

**Comment by BenHenning** [comment]
@rt4914 PTAL for codeowners (I think only your review is left).

**Review by rt4914 — APPROVED** [review]
LGTM, Thanks.

**Pull request merged**
